RecoEgamma/EgammaTools/src/EnergyScaleCorrection.cc

```diff
@@ -0,0 +1,431 @@
+					     unsigned int gainSeed, std::bitset<kErrNrBits> uncBitMask) const
+{
+  
+  const ScaleCorrection* scaleCorr = getScaleCorr(runNumber, et, eta, r9, gainSeed);
+  if(scaleCorr!=nullptr) return scaleCorr->scaleErr(uncBitMask);
+  else return 0.;
+}
+
+
+float EnergyScaleCorrection::smearingSigma(int runnr, double et, double eta, double r9,
+					   unsigned int gainSeed, ParamSmear par, 
+					   float nSigma) const
+{
+  if (par == kRho) return smearingSigma(runnr, et, eta, r9, gainSeed, nSigma, 0.);
+  if (par == kPhi) return smearingSigma(runnr, et, eta, r9, gainSeed, 0., nSigma);
+  return smearingSigma(runnr, et, eta, r9, gainSeed, 0., 0.);
+}
+
+
+float EnergyScaleCorrection::smearingSigma(int runnr, double et, double eta, double r9,
+					   unsigned int gainSeed, float nrSigmaRho, 
+					   float nrSigmaPhi) const
+{
+  const SmearCorrection* smearCorr = getSmearCorr(runnr,et,eta,r9,gainSeed);
+						  
+  if(smearCorr!=nullptr) return smearCorr->sigma(nrSigmaRho,nrSigmaPhi);
+  else return kDefaultSmearVal_;
+}
+
+
+const EnergyScaleCorrection::ScaleCorrection* 
+EnergyScaleCorrection::getScaleCorr(unsigned int runnr, double et, double eta, double r9,
+				    unsigned int gainSeed) const
+{
+
+  // buld the category based on the values of the object
+  CorrectionCategory category(runnr, et, eta, r9, gainSeed);
+  auto result = std::equal_range(scales_.begin(),scales_.end(),category,Sorter<CorrectionCategory,ScaleCorrection>()); 
+  auto nrFound = std::distance(result.first,result.second);
+  if(nrFound==0){
+    LOGWARN("EnergyScaleCorrection") << "[ERROR] Scale category not found: " << category << " Returning uncorrected value.";
+    return nullptr;
+  }else if(nrFound>1){
+    std::ostringstream foundCats;
+    for(auto it = result.first;it!=result.second;++it){
+      foundCats<<"    "<<it->first<<std::endl;
+    }
+    throw cms::Exception("ConfigError") <<" scale error category "<<category<<" has "<<nrFound<<" entries "<<std::endl<<foundCats.str();
+  }  
+  //validate the result, just to be sure
+  if(!result.first->first.inCategory(runnr,et,eta,r9,gainSeed)){
+    throw cms::Exception("LogicError") <<" error found scale category "<<result.first->first<<" that does not contain run "<<runnr<<" et "<<et<<" eta "<<eta<<" r9 "<<r9<<" gain seed "<<gainSeed;
+  }
+  return &result.first->second;
+}
+
+const EnergyScaleCorrection::SmearCorrection* 
+EnergyScaleCorrection::getSmearCorr(unsigned int runnr, double et, double eta, double r9,
+				    unsigned int gainSeed) const
+{
+
+  // buld the category based on the values of the object
+  CorrectionCategory category(runnr, et, eta, r9, gainSeed);
+  auto result = std::equal_range(smearings_.begin(),smearings_.end(),category,Sorter<CorrectionCategory,SmearCorrection>()); 
+  auto nrFound = std::distance(result.first,result.second);
+  if(nrFound==0){
+    LOGWARN("EnergyScaleCorrection") << "[ERROR] Smear category not found: " << category << " Returning uncorrected value.";
```

Comment:
looks like an exception, also

Reply:
same comment as for scaling, it can happen very rarely right now (eg eta of 3.02 which happens). It would be better in the future for it not to happen but right now hands are tied